homeassistant/components/voice_assistant/pipeline.py

```diff
@@ -56,69 +51,165 @@ class Pipeline:
+@dataclass
+class PipelineRun:
+    """Running context for a pipeline."""
+
+    hass: HomeAssistant
+    context: Context
+    pipeline: Pipeline
+    event_callback: Callable[[PipelineEvent], None]
+    language: str = None  # type: ignore[assignment]
+
+    def __post_init__(self):
+        """Set language for pipeline."""
+        self.language = self.pipeline.language or self.hass.config.language
+
+    async def start(self):
+        """Emit run start event."""
+        self.event_callback(
             PipelineEvent(
                 PipelineEventType.RUN_START,
                 {
-                    "pipeline": self.name,
-                    "language": language,
+                    "pipeline": self.pipeline.name,
+                    "language": self.language,
                 },
             )
         )
 
-        intent_input = request.intent_input
+    async def finish(self):
+        """Emit run finish event."""
+        self.event_callback(
+            PipelineEvent(
+                PipelineEventType.RUN_FINISH,
+            )
+        )
 
-        event_callback(
+    async def recognize_intent(
+        self, intent_input: str, conversation_id: str | None
+    ) -> conversation.ConversationResult:
+        """Run intent recognition portion of pipeline."""
+        self.event_callback(
             PipelineEvent(
                 PipelineEventType.INTENT_START,
                 {
-                    "engine": self.conversation_engine or "default",
+                    "engine": self.pipeline.conversation_engine or "default",
                     "intent_input": intent_input,
                 },
             )
         )
 
         conversation_result = await conversation.async_converse(
-            hass=hass,
+            hass=self.hass,
             text=intent_input,
-            conversation_id=request.conversation_id,
-            context=context,
-            language=language,
-            agent_id=self.conversation_engine,
+            conversation_id=conversation_id,
+            context=self.context,
+            language=self.language,
+            agent_id=self.pipeline.conversation_engine,
         )
 
-        event_callback(
+        self.event_callback(
             PipelineEvent(
                 PipelineEventType.INTENT_FINISH,
                 {"intent_output": conversation_result.as_dict()},
             )
         )
 
-        event_callback(
+        return conversation_result
+
+    async def text_to_speech(self, tts_input: str) -> str:
+        """Run text to speech portion of pipeline. Returns URL of TTS audio."""
+        self.event_callback(
             PipelineEvent(
-                PipelineEventType.RUN_FINISH,
+                PipelineEventType.TTS_START,
+                {
+                    "engine": self.pipeline.tts_engine,
+                    "tts_input": tts_input,
+                },
+            )
+        )
+
+        manager: tts.SpeechManager = self.hass.data[tts.DOMAIN]
```

Comment:
You're not allowed to interact directly with objects from other integrations. Only interact with methods they expose on the top-level.

For TTS, we can do:

```python
from homeassistant.components.tts.media_source import generate_media_source_id as tts_generate_media_source_id
from homeassistant.components.media_source import async_resolve_media

url_path = await async_resolve_media(
    hass,
    tts_generate_media_source_id(...),
    None,
)
```